@here/harp-mapview/lib/MapView.ts

```diff
@@ -2574,18 +2571,63 @@ export class MapView extends THREE.EventDispatcher {
         this.m_screenCollisions.update(width, height);
 
         this.m_pixelToWorld = undefined;
+        this.m_fog.update(this, this.m_viewRanges.maximum);
 
+        this.computeTargetAndDistance();
+        this.m_zoomLevel = MapViewUtils.calculateZoomLevelFromDistance(this, this.m_targetDistance);
+    }
+
+    private computeTargetAndDistance() {
         const cameraPitch = MapViewUtils.extractAttitude(this, this.m_camera).pitch;
-        const cameraPosZ = this.getCameraHeightAboveTerrain(TERRAIN_ZOOM_LEVEL);
-        const zoomLevelDistance = cameraPosZ / Math.cos(Math.min(cameraPitch, Math.PI / 3));
-        this.m_zoomLevel = MapViewUtils.calculateZoomLevelFromDistance(this, zoomLevelDistance);
-        this.m_fog.update(this, this.m_viewRanges.maximum);
 
-        const target = MapViewUtils.getWorldTargetFromCamera(this.m_camera, this.projection);
+        // Even for tilt 90 in flat projection getWorldTargetFromCamera is returning some point
+        const target = cameraPitch < PITCH_LIMIT_RAD ? this.raycastTargetFromCamera() : null;
         if (target !== null) {
+            // We have a valid target point. Use it to compute the distance.
             this.m_targetWorldPos.copy(target);
-            this.m_targetGeoPos = this.projection.unprojectPoint(target);
             this.m_targetDistance = this.camera.position.distanceTo(target);
+            this.m_targetGeoPos = this.projection.unprojectPoint(this.m_targetWorldPos);
+        } else {
+            // We either reached the [[PITCH_LIMIT]] or we did not hit the ground surface.
+            // In this case we do the reverse, i.e. compute some fallback distance and
+            // use it to compute the tagret point by using the camera direction.
+            const cameraPosZ = this.getCameraHeightAboveTerrain(TERRAIN_ZOOM_LEVEL);
+
+            //For flat projection we fallback to the target distance at 89 degree pitch.
+            //For spherical projection we fallback to the tangent line distance
+            this.m_targetDistance =
+                this.projection.type === ProjectionType.Planar
+                    ? cameraPosZ / Math.cos(Math.min(cameraPitch, PITCH_LIMIT_RAD))
+                    : Math.sqrt(
+                          Math.pow(cameraPosZ + EarthConstants.EQUATORIAL_RADIUS, 2) -
+                              Math.pow(EarthConstants.EQUATORIAL_RADIUS, 2)
+                      );
+
+            const cameraDir = this.m_camera.getWorldDirection(cache.vector3[0]);
```

Comment:
Shouldn't we use the same fallback (intersection point) for position calculation as for target distance calculus?
Here our target will fly in the void, I would rather assume that we lay back it on the surface:
TargetPos = CameraDir (with tilt constrained) * target distance

Reply:
The whole concept of target point makes not much sense if you have a free camera. We just do this here to always return a target point in our API. We could also change the API to not return a target point at all in these cases but I think also a lot of internal code is relying on a target point.
Currently there is at least one property that always holds for the target point: The (screen)projected target point is always in the middle of the screen. With your proposal that would not be true anymore, but as I said before what we choose does not really matter nobody cares about the target point in free camera mode.
The only important thing is the targetDistance b/c we use it to compute the zoomlevel.